### PR TITLE
upgrade tools

### LINKED ISSUE #3640: nogo_main fail to build when to upgrade golang.org/x/tools
### What version of Bazel are you using?

Build label: 6.3.1-homebrew

### Does this issue reproduce with the latest releases of all the above?

if you update the tools like this

```
    go_repository(
        name = "org_golang_x_tools",
        build_file_proto_mode = "disable_global",
        importpath = "golang.org/x/tools",
        sum = "h1:ojD5zOW8+7dOGzdnNgersm8aPfcDjhMp12UfG93NIMc=",
        version = "v0.11.1",
    )
```

you will get the error

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
external/io_bazel_rules_go/go/tools/builders/nogo_main.go:237:11: not enough arguments in call to pkg.facts.Encode
        have ()
        want (bool)
external/io_bazel_rules_go/go/tools/builders/nogo_main.go:399:54: not enough arguments in call to facts.NewDecoder(pkg.types).Decode
        have (func(pkg *types.Package) ([]byte, error))
        want (bool, func(pkgPath string) ([]byte, error))

```

### PR DESCRIPTION
**Pull request by hawkingrei**
**What type of PR is this?**

> Uncomment one line below and remove others.
>
> Bug fix
> Feature
> Documentation
> Other

**What does this PR do? Why is it needed?**

```x/tools``` is too old to upgrade the linter such as golangci-linter. 

**Which issues(s) does this PR fix?**

Fixes #3640

**Other notes for review**

### CODE REVIEW

**Comment by hawkingrei** [comment]
@sluongng I've decided to push this forward, it seems that it needs to have an upload to mirror?

**Comment by sluongng** [comment]
Yeah use this to request for a mirror https://github.com/bazelbuild/bazel/issues/new?assignees=sgowroji%2Cpavank1992%2Ciancha1992&labels=mirror+request%2Ctype%3A+process%2CP2%2Cteam-OSS&projects=&template=mirror_request.yml&title=%5BMirror%5D+

File: tests/integration/popular_repos/README.rst [hunk 1]
```diff
@@ -191,8 +191,6 @@ This runs tests from the repository `golang.org/x/tools <https://golang.org/x/to
 * @org_golang_x_tools//internal/jsonrpc2:jsonrpc2_test
 * @org_golang_x_tools//internal/jsonrpc2/servertest:servertest_test
 * @org_golang_x_tools//internal/jsonrpc2_v2:jsonrpc2_v2_test
-* @org_golang_x_tools//internal/lockedfile:lockedfile_test
```
Comment (hawkingrei) [inline tests/integration/popular_repos/README.rst:194]: I cannot find this package in the ```x/tools```.

**Comment by hawkingrei** [comment]
@sluongng CI has passed.

**Comment by hawkingrei** [comment]
@fmeum PTAL

**Review by sluongng — APPROVED** [review]
LGTM

Curious readers might want to give https://github.com/golang/tools/releases a read for latest features. A lot of them are `gopls` related, which are irrelevant to rules_go, but the analysis and memory improvements are interesting regardless.

File: go/tools/builders/nogo_main.go [hunk 2]
```diff
@@ -609,8 +609,8 @@ func (i *importer) Import(path string) (*types.Package, error) {
 	return gcexportdata.Read(r, i.fset, i.packageCache, path)
 }
 
-func (i *importer) readFacts(pkg *types.Package) ([]byte, error) {
-	archive := i.factMap[pkg.Path()]
+func (i *importer) readFacts(kgPath string) ([]byte, error) {
```
Comment (tyler-french) [inline go/tools/builders/nogo_main.go:612]: What does `kgPath` mean semantically?

File: go/tools/builders/nogo_main.go [hunk 3]
```diff
@@ -234,7 +234,7 @@ func checkPackage(analyzers []*analysis.Analyzer, packagePath string, packageFil
 
 	// Process diagnostics and encode facts for importers of this package.
 	diagnostics := checkAnalysisResults(roots, pkg)
-	facts := pkg.facts.Encode()
+	facts := pkg.facts.Encode(true)
```
Comment (tyler-french) [inline go/tools/builders/nogo_main.go:237]:
```suggestion
	facts := pkg.facts.Encode(true /* skipMethodSorting */)
```
Could we either attach a meaning to this? I don't want future confusion from API breaking changes to this call.

Naked parameters in function calls can hurt readability.

File: go/tools/builders/nogo_main.go [hunk 4]
```diff
@@ -396,7 +396,7 @@ func load(packagePath string, imp *importer, filenames []string) (*goPackage, er
 	}
 	pkg.types, pkg.typesInfo = types, info
 
-	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(imp.readFacts)
+	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(true, imp.readFacts)
```
Comment (tyler-french) [inline go/tools/builders/nogo_main.go:399]:
```suggestion
	pkg.facts, err = facts.NewDecoder(pkg.types).Decode(true  /* skipMethodSorting */, imp.readFacts)
```

Similar to above. Naked parameters in function calls can hurt readability.

**Review by tyler-french — COMMENTED** [review]
> Thanks for taking this on! Looks good to me.
> 
> @tyler-french Could you give this a test run?

@fmeum working on testing this now. 

@hawkingrei Could you `git rebase -i` and squash this PR into one commit? This might make it a bit easier to test.

Also added a couple of comments.

File: go/tools/builders/nogo_main.go (hunk 2 above)
Comment (hawkingrei) [reply go/tools/builders/nogo_main.go]: It should be ```pkgPath```. This is my fault.

File: go/tools/builders/nogo_main.go (hunk 3 above)
Comment (hawkingrei) [reply go/tools/builders/nogo_main.go]: Added

File: go/tools/builders/nogo_main.go (hunk 4 above)
Comment (hawkingrei) [reply go/tools/builders/nogo_main.go]: Added

**Comment by hawkingrei** [comment]
> > Thanks for taking this on! Looks good to me.
> > @tyler-french Could you give this a test run?
> 
> @fmeum working on testing this now.
> 
> @hawkingrei Could you `git rebase -i` and squash this PR into one commit? This might make it a bit easier to test.
> 
> Also added a couple of comments.

I have rebased it.

**Review by tyler-french — APPROVED** [review]

**Comment by fmeum** [comment]
@hawkingrei Thanks!

**Comment by hawkingrei** [comment]
> @hawkingrei Thanks!

When will we release a new rules_go on the github release ?

**Comment by fmeum** [comment]
@tyler-french is already working on it

**Pull request merged**